src/spi/mod.rs

```diff
@@ -53,3 +53,63 @@ pub const MODE_3: Mode = Mode {
     polarity: Polarity::IdleHigh,
     phase: Phase::CaptureOnSecondTransition,
 };
+
+/// SPI error
+pub trait Error: core::fmt::Debug {
+    /// Convert error to a generic SPI error kind
+    ///
+    /// By using this method, SPI errors freely defined by HAL implementations
+    /// can be converted to a set of generic SPI errors upon which generic
+    /// code can act.
+    fn kind(&self) -> ErrorKind;
+}
+
+/// SPI error kind
+///
+/// This represents a common set of SPI operation errors. HAL implementations are
+/// free to define more specific or additional error types. However, by providing
+/// a mapping to these common SPI errors, generic code can still react to them.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+#[non_exhaustive]
+pub enum ErrorKind {
+    /// An unspecific bus error occurred
+    Bus,
+    /// The peripheral receive buffer was overrun
+    Overrun,
+    /// Multiple devices on the SPI bus are trying across each other, e.g. in a multi-master setup
+    ModeFault,
+    /// CRC does not match the received data
+    Crc,
```

Comment:
If nothing on the traits deals with CRCs, do we want a CRC error?

Reply:
At least STM32s support hardware CRC on SPI. See for example: [AN2576 chapter 3](https://application-notes.digchip.com/005/5-10375.pdf#G1065017)

Reply:
Yeah, but the traits have no way to enable/configure CRC checking in a HAL-independent way. If some driver wants CRC checking, it'll presumably use the HAL-specific stm32xxx-hal APIs to enable it, so it can check for CRC errors using the concrete `Error` type from the HAL instead of through ErrorKind.

IMO the HAL-independent ErrorKind should only have errors relating to the functionalities in HAL-independent traits.

Reply:
If we follow that logic rigorously we can basically remove everything but `Overrun` and `Other` (same thing for serial).
I think there is value in providing variants for most common error conditions, though.

Reply:
I am with @Dirbaio on this one. I say we remove everything but `Overrun` and `Other`, and if a CRC trait or some such is added to the HAL then these sorts of variants can be added at that time.

Reply:
I have now removed `ErrorKind::Crc` as discussed.